pkg/clients/dclient/discovery.go

```diff
@@ -92,88 +93,280 @@ func (c serverPreferredResources) GetGVRFromAPIVersionKind(apiVersion string, ki
+		if groupVersion == "" || kubeutils.GroupVersionMatches(groupVersion, serverResourceList.GroupVersion) {
+			for _, serverResource := range serverResourceList.APIResources {
+				if serverResource.Name == parentResourceName+"/"+strings.ToLower(subresource) {
+					logger.V(4).Info("matched API resource to kind", "apiResource", serverResource, "kind", kind)
+
+					serverResourceGv := getServerResourceGroupVersion(serverResourceList.GroupVersion, serverResource.Group, serverResource.Version)
+					gv, err := schema.ParseGroupVersion(serverResourceGv)
+					if err != nil {
+						logger.Error(err, "failed to parse GV", "groupVersion", serverResourceGv)
+						return nil, schema.GroupVersionResource{}, err
+					}
+
+					serverResource.Group = gv.Group
+					serverResource.Version = gv.Version
+
+					groupVersionResource := gv.WithResource(serverResource.Name)
+					logger.V(4).Info("gv with resource", "gvWithResource", groupVersionResource)
+					return &serverResource, groupVersionResource, nil
+				}
+			}
 		}
+	}
 
-		for _, resource := range serverResource.APIResources {
-			if resourceMatches(resource, kind, subresource) {
-				logger.V(4).Info("matched API resource to kind", "apiResource", resource, "kind", kind)
-				gv, err := schema.ParseGroupVersion(serverResource.GroupVersion)
-				if err != nil {
-					logger.Error(err, "failed to parse GV", "groupVersion", serverResource.GroupVersion)
-					return nil, schema.GroupVersionResource{}, err
+	return nil, schema.GroupVersionResource{}, fmt.Errorf("resource not found for kind %s", kind)
+}
+
+// findResource finds an API resource that matches 'groupVersion', 'kind', in the given serverResourcesList
+func findResource(groupVersion string, kind string, serverPreferredResources, serverGroupsAndResources []*metav1.APIResourceList) (apiResource, parentAPIResource *metav1.APIResource, gvr schema.GroupVersionResource, err error) {
+
+	type APIResourceWithListGV struct {
+		APIResource metav1.APIResource
+		ListGV      string
+	}
+
+	matchingServerResources := make([]APIResourceWithListGV, 0)
+
+	for _, serverResourceList := range serverGroupsAndResources {
+		for _, serverResource := range serverResourceList.APIResources {
+			serverResourceGv := getServerResourceGroupVersion(serverResourceList.GroupVersion, serverResource.Group, serverResource.Version)
+			if groupVersion == "" || kubeutils.GroupVersionMatches(groupVersion, serverResourceGv) {
+				if serverResource.Kind == kind || serverResource.SingularName == kind {
+					gv, err := schema.ParseGroupVersion(serverResourceGv)
+					if err != nil {
+						logger.Error(err, "failed to parse GV", "groupVersion", serverResourceGv)
+						return nil, nil, schema.GroupVersionResource{}, err
+					}
+
+					serverResource.Group = gv.Group
+					serverResource.Version = gv.Version
+
+					matchingServerResources = append(matchingServerResources, APIResourceWithListGV{APIResource: serverResource, ListGV: serverResourceList.GroupVersion})
```

Comment:
I see this duplicated logic in multiple places, can we extract and re-use it?

Reply:
Nested loops and if statement is common in some places but we can't extract a function for them as the body of `if` is different, still I extracted some functions to make the code concise.